Create a getting started with scpca data section

allyhawkins opened the pull request:
Closes #52. 

Here I'm adding the section on how to get started using an Scpca dataset. I've included sections on importing the RDS files into R, quality control, normalization, downstream analysis, and then working with Seurat. 

In the first section I included a short introduction on the two RDS files that are present for each sample and linked out to other sections in the docs when necessary. We do have a FAQ that explains reading into R, but I remembered we talked about repeating some of that information and it seemed like a necessary first step. 

For the quality control section, I included information about using `miQC` and how we have presence of the `miQC_pass` column in the objects. I have some explanation on what `miQC` is along with links to the vignette and documentation. I also inserted code chunks to show how you would filter using that column. 

In addition to using `miQC`, we also have stored the basic metrics in the SCE's that we provide and that's what is recommended in OSCA. I included a section on choosing hard cutoffs and tried to caution people to explore diagnostic plots before simply choosing cutoffs. That's why I didn't provide any recommendations on which cutoffs to use there. Some questions I had about the quality control section were: 

- Does this information seem like too much detail or too little detail for filtering? 
- Should I not include `miQC` or hard cutoffs? 
- Should I include recommended values for hard cutoffs? 

For the normalization and feature selection I generally tried to describe what we would recommend while also linking out to other resources so that they could learn on their own outside of our recommendations. I was back and forth on including code chunks in these sections since I don't want people to blindly follow this as a tutorial, but want people to actually read and stay informed. However, the code chunks seemed useful to have to me so they could follow along for each step? Do others have thoughts on this? 

In terms of the dimensionality reduction section, I included some detail but did not go too far into the weeds of how PCA and UMAP are calculated. I thought that didn't seem too necessary, but let me know if there should be more detail here. 

The last section I included was about using ScPCA data with Seurat. We already have a fairly extensive [FAQ on transferring SCE's to a Seurat object](https://scpca.readthedocs.io/en/latest/faq.html#what-if-i-want-to-use-seurat-instead-of-bioconductor) so I provided a link to that FAQ followed by tutorials on working with Seurat performing the same steps that I described for working with SCE objects. Should there be more explanation here or more code chunks? 

Some other overall questions: 

- Does the organization make sense? Would you divide it into different sections? 
- Is the level of detail and the amount of linking to external resources what you had in mind?

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+As a reminder, each RDS file contains a [`SingleCellExperiment` object](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html). 
+Refer to {ref}`single-cell gene expression file contents<sce_file_contents:single-cell gene expression file contents>` for a more detailed description on the contents of the included `SingleCellExperiment` object. 
+
+In order to work with `SingleCellExperiment` objects in R, we need to ensure that we have the [`SingleCellExperiment` package]() installed and loaded. 
```
This line needs a URL. I assume you intended to include this one.

```suggestion
In order to work with `SingleCellExperiment` objects in R, we need to ensure that we have the [`SingleCellExperiment` package](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) installed and loaded. 
```

jaclyn-taroni commented on docs/getting_started.md:
````diff
@@ -0,0 +1,166 @@
+```r
+# if SingleCellExperiment is not installed, install the package
+# otherwise the installation step can be skipped
+BiocManager::install("SingleCellExperiment")
````
You could consider putting this inside some logic. 

```suggestion
if (!("SingleCellExperiment" %in% installed.packages())) {
  BiocManager::install("SingleCellExperiment")
}
```

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+BiocManager::install("SingleCellExperiment")
+
+library(SingleCellExperiment)
+# read in the RDS file, including the complete path to the file's location
+sce <- readRDS("SCPCL000000_filtered.rds")
```
I wonder if you need to put more information about the path outside of the comment and mention that you're using a placeholder. The way you have the example here assumes that they moved the file outside of the folder structure in the download, right? That seems a little odd to me. I know it's slightly different between project and sample downloads, but there will always be a `sample_id` folder IIRC.

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+
+## Quality control  
+
+After we have imported the RDS file into R and accessed the `SingleCellExperiment` object, we can begin working with the data. 
```
What does "accessed the `SingleCellExperiment` object" mean?

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+## Quality control  
+
+After we have imported the RDS file into R and accessed the `SingleCellExperiment` object, we can begin working with the data. 
+Before we perform any downstream steps, it is recommended to remove any low quality cells from our dataset.
```
```suggestion
Before we perform any downstream steps, we recommend removing low quality cells from the dataset.
```

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+
+After we have imported the RDS file into R and accessed the `SingleCellExperiment` object, we can begin working with the data. 
+Before we perform any downstream steps, it is recommended to remove any low quality cells from our dataset.
+This would include cells that may be dying or damaged, showing a higher percentage of reads coming from mitochondrial cells. 
+Low quality cells may also be present due to technical artifacts such as inefficient reverse transcription or PCR amplification resulting in a lower number of total reads and unique genes identified. 
```
Don't need to use this exact wording, but it's helpful to start with _what you're filtering out_ and then say _why_ in my opinion.

```suggestion
Low quality cells include those with a higher percentage of reads from mitochondrial genes (i.e., those that are damaged or dying) and those with a lower number of total reads and unique genes identified (i.e., those with inefficient reverse transcription or PCR amplification). 
```

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+This would include cells that may be dying or damaged, showing a higher percentage of reads coming from mitochondrial cells. 
+Low quality cells may also be present due to technical artifacts such as inefficient reverse transcription or PCR amplification resulting in a lower number of total reads and unique genes identified. 
+
+For the filtered `SingleCellExperiment` object present in all `filtered.rds` files, we have used [`miQC`](https://bioconductor.org/packages/release/bioc/html/miQC.html), a data driven approach to predict low-quality cells, to identify cells that should be removed prior to downstream analyses. 
```
```suggestion
All `filtered.rds` objects include [`miQC`](https://bioconductor.org/packages/release/bioc/html/miQC.html) results that can be used to identify cells that should be removed prior to downstream analyses. 
```

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+Low quality cells may also be present due to technical artifacts such as inefficient reverse transcription or PCR amplification resulting in a lower number of total reads and unique genes identified. 
+
+For the filtered `SingleCellExperiment` object present in all `filtered.rds` files, we have used [`miQC`](https://bioconductor.org/packages/release/bioc/html/miQC.html), a data driven approach to predict low-quality cells, to identify cells that should be removed prior to downstream analyses. 
+`miQC` jointly models the proportion of mitochondrial reads and the number of unique genes detected in each cell to calculate the probability of a cell being compromised (i.e. dead or damaged). 
```
```suggestion
`miQC` jointly models the proportion of mitochondrial reads and the number of unique genes detected in each cell to calculate the probability of a cell being compromised (i.e., dead or damaged). 
```

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+For the filtered `SingleCellExperiment` object present in all `filtered.rds` files, we have used [`miQC`](https://bioconductor.org/packages/release/bioc/html/miQC.html), a data driven approach to predict low-quality cells, to identify cells that should be removed prior to downstream analyses. 
+`miQC` jointly models the proportion of mitochondrial reads and the number of unique genes detected in each cell to calculate the probability of a cell being compromised (i.e. dead or damaged). 
+High-quality cells are those with a low probability of being being compromised (< 0.75) or sufficiently low mitochondrial content. 
+We have already incorporated the `miQC` results into the {ref}`colData() of the SingleCellExperiment object<sce_file_contents:cell metrics>`. 
```
I would maybe move this link to the `colData()` section to the sentence in my first suggestion.

jaclyn-taroni commented on docs/getting_started.md:
````diff
@@ -0,0 +1,166 @@
+
+# filter the `SingleCellExperiment`
+filtered_sce <- sce[, cells_to_keep]
+```
+For more information on using `miQC` for filtering cells, see the following resources: 
````
Can you put a line break between these two lines please? For readability 👀

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+- [Hippen _et al._ 2021](https://doi.org/10.1371/journal.pcbi.1009290)
+
+Alternatively, you can directly filter cells based on having a minimum number of unique genes, total reads, and a maximum fraction of mitochondrial reads. 
+Using the function, [`scuttle::addPerCellQCMetrics()`](https://rdrr.io/github/LTLA/scuttle/man/addPerCellQCMetrics.html), we can calculate the total RNA counts, unique genes, and mitochondrial fraction for each cell and store the results in the `colData()`. 
+These metrics can be used to directly filter the `SingleCellExperiment` object based on informed minimum thresholds. 
+The `filtered.rds` files contain `SingleCellExperiment` objects with these metrics added to the `colData()` and you can read more about all metrics that are included in the {ref}`cell metrics section of the SingleCellExperiment file contents<sce_file_contents:cell metrics>`.
```
So these results are already in the object right? (I am under the impression that you need to add the QC metrics _before_ you run `miQC`.) Start with the fact that these are already present and you can use them to filter.

jaclyn-taroni commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+
+## Normalization 
+
+Single-cell RNA sequencing inherently comes with technical confounders that, if not accounted for, can impact the interpretation of downstream analysis. 
+Technical variables include differences in sequencing depth and cell size across cells in a single-cell library ([Stegle _et al._ 2015](https://doi.org/10.1038/nrg3833).)
+
+If we were to proceed with downstream analysis such as differential gene expression and cell clustering without accounting for such effects, these technical problems could impact the overall conclusions.
+We need to normalize our data to account for the technical variation between cells in our data and maximize biological variation, allowing us to make more accurate conclusions from downstream analysis.
```
I'm torn about whether we need this much information here or if it's better to get right into what we recommend. We do provide links out to other resources.

jaclyn-taroni replied:
> I tend to think this document should have less rather than more

I agree with this take and would extend it to maybe mean cutting the downstream analysis portion.

allyhawkins replied:
To clarify, do you mean completely remove the downstream analysis portion, or to just simplify it?

jaclyn-taroni replied:
I meant completely remove, but after taking a look at @jashapiro's comments I propose: you start by responding to his comments (e.g., removing feature selection) and then we'll see if we want to cut further.

jaclyn-taroni reviewed (commented):
I'm returning some comments on the sections that I think are the minimum viable getting started page. I'm not sure if I expected to see the downstream analysis section here at this point.

jashapiro reviewed (commented):
I had a number of smaller comments about links and wording and such. The overall comment is probably more about my thoughts about how to keep this focused as a "getting started" with a few simple examples and less "theory." I want people to know all of these things, but I am not sure this document is the place to teach them...

allyhawkins commented:
I went through and incorporated the suggestions from both @jaclyn-taroni and @jashapiro and made the following changes: 

- I added the columns description for the `colData` (taken from the sce contents section), stating that these columns are added by `scuttle::addPerCellQCMetrics` and can be found in the `colData`. @jashapiro is this what you had in mind for your comment about adding the names of the columns in? 
- I removed the background information in the normalization section, simply stating that the counts were unnormalized and needed to be normalized, provided the code to do so and then some links to more information 
- I removed the section on variable features 
- I simplified the dimensionality reduction section to have less intro about PCA and UMAP and more here is how to do them. However, I did keep in the cautionary statements about interpreting UMAP since Josh had mentioned he liked having them there. In this section I added a note about selecting variable features and linked to the chapter in OSCA. Is this more what you had in mind here or do you want me to remove more description? 
- I added a section on scanpy. I tried to mirror the section on Seurat that I had and link out to the FAQ and then provide a few links. If there are other links that you think would be helpful please let me know as the only reliable ones I had were from the scanpy tutorials and there isn't much there.

jashapiro commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+High-quality cells are those with a low probability of being being compromised (< 0.75) or sufficiently low mitochondrial content.  
+All cells that are identified as low-quality cells will have `FALSE` in the `miQC_pass` column of the `colData()` and can be removed prior to downstream analyses. 
+
+The following command can be used to remove those cells: 
```
A small clarification
```suggestion
The following command can be used to the low-quality cells: 
```

jashapiro commented on docs/getting_started.md:
```diff
@@ -0,0 +1,166 @@
+| `subsets_mito_percent`  | Percent of all UMI counts assigned to mitochondrial genes                                                                                                                                     |
+| `total`                 | Total UMI count for RNA-seq data and any alternative experiments (i.e., CITE-seq)                                                                                                             |
+
+These metrics can be used to directly filter the `SingleCellExperiment` object based on informed minimum thresholds. 
+
+If you are planning to filter low quality cells using these hard cutoffs, we encourage you to read more about the various metrics and plot the distribution of each metric before deciding on which cells to exclude. 
```
Seemed like this could all be one paragraph. I also removed `minimum` since for at least one people are likely to use a max. 

```suggestion
These metrics can be used to directly filter the `SingleCellExperiment` object based on informed thresholds. 
If you are planning to filter low quality cells using such thresholds, we encourage you to read more about the various metrics and plot the distribution of each metric before deciding on which cells to exclude. 
```

allyhawkins commented:
> I might reduce the UMAP caution section to just the crux, something like: "Use caution when interpreting UMAP results; the location, spacing, and density of the clusters can be dependent on parameter choices and random effects, and may not accurately reflect relationships among cells." I mostly want to get it down to 1-2 sentences, as a reminder of things they have hopefully heard before.

I went ahead and incorporated this suggestion and paired it down to 1 sentence along with incorporating the other minor suggestions. Let me know if this looks as expected.

jashapiro commented on docs/getting_started.md:
````diff
@@ -0,0 +1,164 @@
+
+```r
+# Run UMAP using already stored PCA results
+normalized_sce <- scater::runUMAP(normalized_sce, 
+                          dimred = "PCA")
````
Spacing was off, but I think we can do one line here anyway
```suggestion
normalized_sce <- scater::runUMAP(normalized_sce, dimred = "PCA")
```

jashapiro reviewed (approved):
LGTM, with one little code format update.
